### ISSUE TITLE
[PT2.0][compile] embedding bag  max_norm=-1 causes leaf Variable that requires grad is being used in an in-place operation

### ISSUE THREAD

**Issue by jay746** [pytorch/pytorch#95167]
### 🐛 Describe the bug

Please use below code to reproduce:
```
import torch
class BasicModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, input_dict):
        return torch.nn.functional.embedding_bag(**input_dict)


if __name__ == "__main__":
    input_dict = {
        "input": torch.LongTensor(
            [3, 1, 1, 5, 4, 3, 2, 6, 3, 5, 6, 5, 6, 0, 6, 4, 3, 5, 7, 7]
        ),
        "weight": torch.tensor(
            [
                [0.4597, 0.3764],
                [0.0013, 0.1590],
                [0.3964, 0.0136],
                [0.2254, 0.2905],
                [0.8277, 0.8274],
                [0.6265, 0.7479],
                [0.1626, 0.6431],
                [0.1259, 0.1363],
            ],
            requires_grad=True,
        ),
        "offsets": torch.LongTensor([0, 3, 10]),
        "mode": "mean",
        "sparse": False,
        "scale_grad_by_freq": False,
        "max_norm": -1.0,
    }

    model = BasicModule()
    model = torch.compile(model)
    fwd_res = model(input_dict)
    print(fwd_res)
```

if we comment max_norm then it works. with max_norm same graph works fine in eager exeuction

### Versions

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] torch==2.0.0.dev20230209+cpu
[pip3] torchaudio==2.0.0.dev20230209+cpu
[pip3] torchvision==0.15.0.dev20230209+cpu

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by bdhirsh** [comment]
this looks like an issue around aot autograd seeing mutations and grad / no_grad mode

**Comment by bdhirsh** [comment]
yep, here's a min repro:
```
import torch

@torch.compile(backend="aot_eager")
def f(x):
    x.detach().add_(1)
    return x * 2

out = f(torch.ones(2, requires_grad=True))
```

The problem is that this is an input mutation where the original input required gradients, but the view did not (since it was detached).

We definitely want to fix this - I need to think about the right way to do this (I remember some discussion with Ed and Alban about it a while ago, but I can't find the previous discussion).

**Comment by ezyang** [comment]
I think we discussed this:

```
x.mul_(2).detach().mul_(3)
```

and you'd be obligated to do gradients for the x2, but NOT for the x3. But I think we concluded in the end that everything would work out, because you would copy_ into x the CUSTOM gradient (that would only have x2 but not x3).

So this suggests that we just need to avoid the copy if there were no gradient carrying mutations. Actually a very cheap way to check this would just be to peek if x is a leaf. If it is, then it's impossible to have done a gradient carrying mutation on it. If x is not a leaf, copy_'ing into it is harmless (it's gradient would always be zero), although it's marginally better not to copy so it may still be better to keep track if you've actually mutated an input through an autograd-carrying view.

**Comment by ezyang** [comment]
But let's do the easy thing first.

**Comment by bdhirsh** [comment]
Hmm there are two things here that I think need clarification:


> So this suggests that we just need to avoid the copy if there were no gradient carrying mutations

(1) We don't want to avoid the copy - right? We just want to perform the copy, but hide it from autograd.

 One option would be to perform the copy_() under `torch.no_grad()`. That seems mostly right - but not completely. One case where the behaviors diverge is that mutating a tensor under no_grad() will update its version counter. But taking a detach'd view of a tensor and mutating it will not update the original tensor's version counter.
 
 That makes me think that in this case, we'd just want to do something like "if mutated input is a leaf, run `inpt.detach().copy_()` instead of `inpt.copy_()`. But lmk what you think
 
(2) ~~There's also a more general version of this problem~~
```
def f(x):
    x.mul_(2)
    with torch.no_grad():
        x.mul_(3)
    x.mul_(4)
```

Nvm, I confirmed locally and random blocks of `torch.no_grad()` in the user code involving input mutations actually works fine. That's probably because we trace all-in-one-go, so we've already finished running autograd by the time we get out our ATen graphs and the no_grad() regions have disappeared.

This case is broken today, because `torch.no_grad()` evaporates
